File: python/bullmq/worker.py
from typing import Callable
from uuid import uuid4
from redis.exceptions import (
    BusyLoadingError,
    ConnectionError as RedisConnectionError,
    TimeoutError as RedisTimeoutError,
)
from bullmq.custom_errors import UnrecoverableError, WaitingChildrenError
from bullmq.scripts import Scripts
from bullmq.redis_connection import RedisConnection
from bullmq.event_emitter import EventEmitter
from bullmq.job import Job
from bullmq.timer import Timer
from bullmq.types import WorkerOptions
from bullmq.utils import extract_result

import asyncio
import errno
import re
import traceback
import time
import math

maximum_block_timeout = 10
# 1 millisecond is chosen because the granularity of our timestamps are milliseconds.
# Obviously we can still process much faster than 1 job per millisecond but delays and
# rate limits will never work with more accuracy than 1ms.
minimum_block_timeout = 0.001

# Short delay (in seconds) used to space out retries after a transient error
# in the main worker loop. Mirrors the DELAY_TIME_1 constant used by the
# Node.js implementation for the same purpose.
short_retry_delay = 0.1

# Errnos that indicate a transient/retryable network failure. Used by
# Worker.isConnectionError to classify bare OSErrors raised before the
# redis client has a chance to wrap them.
TRANSIENT_ERRNOS = {
    getattr(errno, name)
    for name in ("ECONNREFUSED", "ECONNRESET", "ENETUNREACH",
                 "EHOSTUNREACH", "EPIPE")
    if hasattr(errno, name)
}

# asyncio aggregates per-host connect attempts into a single OSError
# whose message looks like "Multiple exceptions: [Errno 61] Connect call
# failed ('127.0.0.1', 6379)". In that case error.errno is None, so we
# extract the embedded errno from the message instead.
_ERRNO_PATTERN = re.compile(r"\[Errno (\d+)\]")


class Worker(EventEmitter):
    def __init__(self, name: str, processor: Callable[[Job, str], asyncio.Future], opts: WorkerOptions = {}):
        super().__init__()
        self.name = name
        self.processor = processor
        final_opts = {
            "drainDelay": 5,
            "concurrency": 1,
            "lockDuration": 30000,
            "maxStalledCount": 1,
            "stalledInterval": 30000,
            "runRetryDelay": 15000,
        }
        final_opts.update(opts or {})
        self.opts = final_opts
        redis_opts = opts.get("connection", {})
        self.redisConnection = RedisConnection(redis_opts)
        self.blockingRedisConnection = RedisConnection(redis_opts)
        self.client = self.redisConnection.conn
        self.bclient = self.blockingRedisConnection.conn
        self.prefix = opts.get("prefix", "bull")
        self.scripts = Scripts(opts.get("prefix", "bull"), name, self.redisConnection)
        self.closing = False
        self.forceClosing = False
        self.closed = False
        self.running = False
        self.paused = False
        self.processing = set()
        self.jobs = set()
        self.id = uuid4().hex
        self.waiting = None
        self.blockUntil = 0
        self.limitUntil = 0
        self.drained = False
        self.qualifiedName = self.scripts.queue_keys.getQueueQualifiedName(name)
        self.workerName = opts.get("name")
        self.clientName = self.qualifiedName + (f":w:{self.workerName}" if self.workerName else "")
        self._client_name_set = False

        if processor:
            if opts.get("autorun", True):
                asyncio.ensure_future(self.run())

    async def run(self):
        if self.running:
            raise Exception("Worker is already running")

        await self._ensure_client_names()

        self.timer = Timer(
            (self.opts.get("lockDuration") / 2) / 1000, self.extendLocks, self.emit)
        self.stalledCheckTimer = Timer(self.opts.get(
            "stalledInterval") / 1000, self.runStalledJobsCheck, self.emit)
        self.running = True
        jobs = []

        token_postfix = 0

        while not self.closed:
            while not self.waiting and len(self.processing) < self.opts.get("concurrency") and not self.closing:
                token_postfix+=1
                token = f'{self.id}:{token_postfix}'
                
                # Use retryIfFailed to wrap getNextJob call, similar to TypeScript worker
                async def get_next_job_wrapped():
                    return await self.getNextJob(token)
                
                waiting_job = asyncio.ensure_future(
                    self.retryIfFailed(
                        get_next_job_wrapped,
                        {
                            "delay_in_ms": self.opts.get("runRetryDelay"),
                            "only_emit_error": True,
                        }
                    )
                )
                self.processing.add(waiting_job)

            try:
                jobs, pending = await getCompleted(self.processing, self.emit)

                jobs_to_process = [self.processJob(job, job.token) for job in jobs]
                processing_jobs = [asyncio.ensure_future(
                    j) for j in jobs_to_process]
                pending.update(processing_jobs)
                self.processing = pending

                if (len(jobs) == 0 or len(self.processing) == 0) and self.closing:
                    # We are done processing so we can close the queue
                    break

            except Exception as e:
                # This should never happen or we will have an endless loop
                traceback.print_exc()
                return

        self.running = False
        self.timer.stop()
        self.stalledCheckTimer.stop()

    async def getNextJob(self, token: str):
        """
        Returns a promise that resolves to the next job in queue.
        @param token: worker token to be assigned to retrieved job
        @returns a Job or undefined if no job was available in the queue.
        """
        await self._ensure_client_names()
        job_instance = None
        if not self.waiting and self.drained:
            self.waiting = self.waitForJob()

            try:
                self.blockUntil = await self.waiting
                timestamp = int(time.time() * 1000)

                if self.blockUntil <= 0 or self.blockUntil <= timestamp:
                    job_instance = await self.moveToActive(token)
            finally:
                self.waiting = None
        else:
            job_instance = await self.moveToActive(token)

        if job_instance:
            self.emit("active", job_instance, "waiting")
            return job_instance

    async def moveToActive(self, token: str):
        result = await self.scripts.moveToActive(token, self.opts)
        job_data = None
        id = None
        limit_until = None
        delay_until = None

        if result:
            job_data, id, limit_until, delay_until = result

        return self.nextJobFromJobData(job_data, id, limit_until, delay_until, token)

    def nextJobFromJobData(self, job_data: dict | None = None, job_id: str | None = None, limit_until: int = 0,
        delay_until: int = 0, token: str | None = None) -> Job | None:
        self.limitUntil = max(limit_until, 0) or 0

        if not job_data:
            if not self.drained:
                self.drained = True
                self.blockUntil = 0

        if delay_until:
            self.blockUntil = max(int(delay_until), 0) or 0

        if job_data:
            self.drained = False
            job_instance = Job.fromJSON(self, job_data, job_id)
            job_instance.token = token
            return job_instance

    async def waitForJob(self) -> int:
        block_timeout = self.getBlockTimeout(self.blockUntil)
        block_timeout = block_timeout if self.blockingRedisConnection.capabilities.get("canDoubleTimeout", False) else math.ceil(block_timeout)

        try:
            result = await self.bclient.bzpopmin(self.scripts.keys["marker"], block_timeout)
        except Exception:
            # Short-sleep before re-raising so the outer retryIfFailed loop
            # cannot spin at full CPU while Redis is unavailable. The error
            # itself is emitted exactly once by retryIfFailed when it
            # propagates up the run() call chain.
            if not self.closing and not self.closed:
                await asyncio.sleep(short_retry_delay)
            raise
        if result:
            [_key, member, score] = result

            if member:
                return int(score)
            else:
                return 0
        return 0

    async def _ensure_client_names(self):
        if self._client_name_set:
            return

        await self.redisConnection.set_client_name(self.clientName)
        await self.blockingRedisConnection.set_client_name(self.clientName)
        self._client_name_set = True

    def getBlockTimeout(self, block_until: int) -> float:
        if block_until:
            block_timeout = None
            block_delay = block_until - int(time.time() * 1000)
            if block_delay < self.minimumBlockTimeout * 1000:
                return self.minimumBlockTimeout
            else:
                block_timeout = block_delay / 1000
            # We restrict the maximum block timeout to 10 second to avoid
            # blocking the connection for too long in the case of reconnections
            # reference: https://github.com/taskforcesh/bullmq/issues/1658
            return min(block_timeout, maximum_block_timeout)
        else:
            return max(self.opts.get("drainDelay", 5), self.minimumBlockTimeout)

    @property
    def minimumBlockTimeout(self):
        return minimum_block_timeout if self.blockingRedisConnection.capabilities.get("canBlockFor1Ms", True) else 0.002

    async def processJob(self, job: Job, token: str):
        try:
            # Set worker-level remove options on job if not already set
            if "removeOnComplete" not in job.opts and "removeOnComplete" in self.opts:
                job.opts["removeOnComplete"] = self.opts["removeOnComplete"]
            if "removeOnFail" not in job.opts and "removeOnFail" in self.opts:
                job.opts["removeOnFail"] = self.opts["removeOnFail"]

            self.jobs.add((job, token))
            
            if job.deferredFailure:
                await job.moveToFailed(UnrecoverableError(job.deferredFailure), token)
                self.emit("failed", job, UnrecoverableError(job.deferredFailure))
                return

            result = await self.processor(job, token)
            if not self.forceClosing:
                await self.scripts.moveToCompleted(job, result, job.opts.get("removeOnComplete", False), token, fetchNext=False)
                job.returnvalue = result
                job.attemptsMade = job.attemptsMade + 1
            self.emit("completed", job, result)
        except WaitingChildrenError:
            return
        except Exception as err:
            try:
                if not self.forceClosing:
                    await job.moveToFailed(err, token)

                self.emit("failed", job, err)
            except Exception as err:
                self.emit("error", err, job)
        finally:
            self.jobs.discard((job, token))

    async def retryIfFailed(self, fn, opts=None):
        """
        Retry a coroutine function if it fails, with delay and max retries.
        :param fn: Coroutine function to execute.
        :param opts: Dictionary with options:
            - delay_in_ms: Delay between retries in milliseconds.
            - max_retries: Maximum number of retries.
            - only_emit_error: If True, only emit error and do not raise.
        """
        if opts is None:
            opts = {}
        delay_in_ms = opts.get("delay_in_ms", 15000)
        max_retries = opts.get("max_retries", float('inf'))
        only_emit_error = opts.get("only_emit_error", False)

        retry = 0
        while retry < max_retries:
            try:
                return await fn()
            except Exception as err:
                # Check if this is a connection error that should be retried
                is_connection_error = self.isConnectionError(err)

                if not is_connection_error:
                    # Swallow error if locally not paused or not closing since we did not force a disconnection
                    if not (self.paused or self.closing):
                        self.emit("error", err)

                    if only_emit_error:
                        # Without a short backoff a non-connection error
                        # that keeps happening (e.g. a misconfigured script
                        # or a transient Redis error type we don't classify
                        # as a connection error yet) would cause the outer
                        # worker loop to re-enter retryIfFailed immediately
                        # and busy-loop the CPU. Sleep briefly here so the
                        # retry rate stays bounded. See issue #3103.
                        if not self.closing and not self.closed:
                            await asyncio.sleep(short_retry_delay)
                        return None
                    else:
                        raise err
                else:
                    # For connection errors, wait and retry
                    if delay_in_ms and not self.closing and not self.closed:
                        await asyncio.sleep(delay_in_ms / 1000.0)

                    retry += 1
                    if retry >= max_retries:
                        # If we've reached max retries, raise the last error
                        raise err

        return None

    def isConnectionError(self, error: Exception) -> bool:
        """
        Check if an error is a connection-related error that should trigger a
        retry with backoff rather than being surfaced to the user.

        This mirrors the Node.js worker's `isNotConnectionError` check. We
        deliberately classify a broad range of transient failures as
        connection errors so that losing the Redis server does not cause the
        worker to busy-loop (see issue #3103). In particular, Redis client
        libraries can surface a dropped connection as `ConnectionError`,
        `TimeoutError`, `BusyLoadingError`, or even a bare `OSError` /
        `asyncio.TimeoutError` depending on where the failure occurs.
        """
        if isinstance(
            error,
            (
                RedisConnectionError,
                RedisTimeoutError,
                BusyLoadingError,
                ConnectionRefusedError,
                ConnectionResetError,
                asyncio.TimeoutError,
            ),
        ):
            return True

        # DNS or socket failures raised before the redis client has a
        # chance to wrap them surface as a plain OSError. Match either
        # error.errno or any [Errno N] embedded in the message string.
        if isinstance(error, OSError):
            if error.errno in TRANSIENT_ERRNOS:
                return True
            return any(
                int(n) in TRANSIENT_ERRNOS
                for n in _ERRNO_PATTERN.findall(str(error))
            )

        return False

    async def extendLocks(self):
        # Renew all the locks for the jobs that are still active
        try:
            multi = self.client.pipeline()
            for job, token in self.jobs:
                await self.scripts.extendLock(job.id, token, self.opts.get("lockDuration"), multi)
            result = await multi.execute()

            # result includes an object with locks that may not have been renewed.
            # We should emit an error for each of those jobs.
            #    for jobId, err in result.items():
            #    self.emit("error", "could not renew lock for job " + jobId)

        except Exception as e:
            traceback.print_exc()

    async def runStalledJobsCheck(self):
        try:
            stalled = await self.scripts.moveStalledJobsToWait(self.opts.get("maxStalledCount"), self.opts.get("stalledInterval"))
            for jobId in stalled:
                self.emit("stalled", jobId)

        except Exception as e:
            self.emit('error', e)

    async def close(self, force: bool = False):
        """
        Closes the worker and related redis connections.

        This method waits for current jobs to finalize before returning.
        """
        self.closing = True
        if force:
            self.forceClosing = True
            self.cancelProcessing()

        if not force and len(self.processing) > 0:
            await asyncio.wait(self.processing, return_when=asyncio.ALL_COMPLETED)

        for conn in (self.blockingRedisConnection, self.redisConnection):
            try:
                await conn.close()
            except Exception as err:
                self.emit('error', err)

        self.closed = True
        self.emit('closed')

    async def pause(self, do_not_wait_active: bool = False):
        """
        Pauses the worker, preventing it from processing new jobs.

        This method waits for current jobs to finalize before returning.
        """
        if not self.paused:
            self.paused = True
            if not do_not_wait_active and len(self.processing) > 0:
                await asyncio.wait(self.processing, return_when=asyncio.ALL_COMPLETED)
        self.emit('paused')

    def resume(self):
        """
        Resumes processing of this worker (if paused).
        """
        if self.paused:
            self.paused = False
            self.emit('resumed')

    def cancelProcessing(self):
        for job in self.processing:
            if not job.done():
                job.cancel()


async def getCompleted(task_set: set, emit_callback) -> tuple[list[Job], set]:
    job_set, pending = await asyncio.wait(task_set, return_when=asyncio.FIRST_COMPLETED)
    jobs = [extract_result(job_task, emit_callback) for job_task in job_set]
    # we filter `None` out to remove:
    # a) an empty 'completed jobs' list; and
    # b) a failed extract_result
    jobs = list(filter(lambda j: j is not None, jobs))
    return jobs, pending

File: dotnet/src/BullMQ/Worker.cs
namespace BullMQ;

/// <summary>
/// A processor function that handles a job and returns its result. The returned
/// value is stored as the job's return value on completion.
/// </summary>
public delegate Task<object?> Processor(Job job, CancellationToken cancellationToken);

/// <summary>
/// Processes jobs from a queue. As soon as it is created (and unless
/// <see cref="WorkerOptions.Autorun"/> is disabled) it starts fetching and
/// processing jobs concurrently up to <see cref="WorkerOptions.Concurrency"/>.
///
/// <para>
/// Jobs are fetched using the backend's blocking wait primitive
/// (Redis <c>BZPOPMIN</c> on the marker set / PostgreSQL <c>LISTEN</c>), so an
/// idle worker does not busy-poll. Locks are renewed periodically by a
/// <see cref="LockManager"/>, and a background stalled-job checker recovers jobs
/// whose lock expired.
/// </para>
/// </summary>
public sealed class Worker : IAsyncDisposable
{
    private readonly Lazy<Task<IQueueBackend>> _backend;
    private readonly Processor _processor;

    // Cancels the processor's abort signal (force close only).
    private readonly CancellationTokenSource _abortCts = new();

    // Interrupts the fetcher's waits (any close).
    private readonly CancellationTokenSource _closeCts = new();

    private SemaphoreSlim? _slots;
    private long _tokenCounter;
    private volatile bool _closing;
    private volatile bool _forceClosing;
    private volatile bool _drained;
    private Task? _running;
    private LockManager? _lockManager;
    private Task? _stalledChecker;

    /// <summary>A unique id for this worker instance.</summary>
    public string Id { get; } = Guid.NewGuid().ToString();

    /// <summary>The queue name this worker consumes from.</summary>
    public string Name { get; }

    /// <summary>The options this worker was created with.</summary>
    public WorkerOptions Opts { get; }

    /// <summary>Raised when a job becomes active (starts processing).</summary>
    public event Action<Job>? Active;

    /// <summary>Raised when a job completes successfully.</summary>
    public event Action<Job, object?>? Completed;

    /// <summary>Raised when a job fails.</summary>
    public event Action<Job?, Exception>? Failed;

    /// <summary>Raised when the worker drains the waiting list.</summary>
    public event Action? Drained;

    /// <summary>Raised when a non-fatal error occurs (e.g. a fetch failure).</summary>
    public event Action<Exception>? Error;

    /// <summary>Raised when a job has stalled and been moved back to wait.</summary>
    public event Action<string>? Stalled;

    /// <summary>Raised when lock renewal fails for one or more jobs.</summary>
    public event Action<IReadOnlyList<string>>? LockRenewalFailed;

    /// <summary>Raised when locks are successfully renewed.</summary>
    public event Action<IReadOnlyList<string>>? LocksRenewed;

    public Worker(string name, Processor processor, WorkerOptions opts)
    {
        Name = name;
        _processor = processor ?? throw new ArgumentNullException(nameof(processor));
        Opts = opts ?? throw new ArgumentNullException(nameof(opts));

        if (opts.Postgres is null && opts.Connection is null)
        {
            throw new ArgumentException("Worker requires a connection");
        }

        _backend = new Lazy<Task<IQueueBackend>>(InitBackendAsync);

        if (opts.Autorun)
        {
            _running = RunAsync();
        }
    }

    private async Task<IQueueBackend> InitBackendAsync()
    {
        var backend = await BackendBuilder
            .CreateAsync(Name, Opts, Opts.LockDuration, Opts.Name, withBlockingConnection: true)
            .ConfigureAwait(false);
        await backend.WaitUntilReadyAsync().ConfigureAwait(false);

        var lockRenewTime = Opts.LockRenewTime ?? Opts.LockDuration / 2;
        _lockManager = new LockManager(
            backend,
            lockRenewTime,
            Opts.LockDuration,
            err => Error?.Invoke(err),
            ids => LockRenewalFailed?.Invoke(ids),
            ids => LocksRenewed?.Invoke(ids));

        return backend;
    }

    /// <summary>Waits until the worker's connection is ready.</summary>
    public async Task WaitUntilReadyAsync() => await _backend.Value.ConfigureAwait(false);

    /// <summary>
    /// Starts processing (only needed when <see cref="WorkerOptions.Autorun"/> is
    /// disabled). Returns a task that completes when the worker stops.
    /// </summary>
    public Task RunAsync()
    {
        _running ??= RunInternalAsync();
        return _running;
    }

    private async Task RunInternalAsync()
    {
        var backend = await _backend.Value.ConfigureAwait(false);
        _slots = new SemaphoreSlim(Opts.Concurrency, Opts.Concurrency);

        if (!Opts.SkipLockRenewal)
        {
            _lockManager?.Start();
        }

        if (!Opts.SkipStalledCheck)
        {
            _stalledChecker = StalledCheckerLoopAsync(backend);
        }

        await MainLoopAsync(backend).ConfigureAwait(false);
    }

    /// <summary>
    /// The main fetch loop. A semaphore bounds in-flight processing to
    /// <see cref="WorkerOptions.Concurrency"/>: each free slot triggers a fetch,
    /// jobs are processed fire-and-forget (releasing their slot on completion),
    /// and an idle worker blocks on the backend's wait primitive instead of
    /// busy-polling.
    /// </summary>
    private async Task MainLoopAsync(IQueueBackend backend)
    {
        while (!_closing)
        {
            try
            {
                await _slots!.WaitAsync(_closeCts.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                break;
            }

            if (_closing)
            {
                _slots!.Release();
                break;
            }

            var token = $"{Id}:{Interlocked.Increment(ref _tokenCounter)}";
            NextJobData next;
            try
            {
                next = await backend.MoveToActiveAsync(token, Opts.Name).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                _slots!.Release();
                if (!_closing)
                {
                    Error?.Invoke(ex);
                }

                await SafeDelayAsync(200).ConfigureAwait(false);
                continue;
            }

            if (next.Job is not null)
            {
                _drained = false;
                var job = Job.FromJson(backend, Name, next.Job, next.JobId);
                job.Token = token;
                if (!string.IsNullOrEmpty(job.RepeatJobKey))
                {
                    await ScheduleNextIterationAsync(backend, job).ConfigureAwait(false);
                }

                _ = RunJobAsync(backend, job, token);
            }
            else
            {
                if (!_drained)
                {
                    _drained = true;
                    Drained?.Invoke();
                }

                _slots!.Release();
                await WaitForJobAsync(backend, next.DelayUntil).ConfigureAwait(false);
            }
        }

        // Drain: wait for every in-flight job to finish (skipped on force close).
        if (!_forceClosing)
        {
            for (var i = 0; i < Opts.Concurrency; i++)
            {
                await _slots!.WaitAsync().ConfigureAwait(false);
            }
        }
    }

    private async Task RunJobAsync(IQueueBackend backend, Job job, string token)
    {
        try
        {
            await ProcessJob(backend, job, token).ConfigureAwait(false);
        }
        finally
        {
            _slots!.Release();
        }
    }

    /// <summary>
    /// When a fetched job was produced by a scheduler, enqueue the scheduler's
    /// next iteration (matching the reference worker's reschedule-on-fetch).
    /// </summary>
    private async Task ScheduleNextIterationAsync(IQueueBackend backend, Job job)
    {
        try
        {
            if (job.Repeat is null)
            {
                return;
            }

            if (!await backend.IsJobSchedulerAsync(job.RepeatJobKey!).ConfigureAwait(false))
            {
                return;
            }

            var scheduler = new JobScheduler(backend, Name);
            await scheduler.UpsertJobSchedulerAsync(
                job.RepeatJobKey!, job.Repeat, job.Name, job.Data, job.Opts,
                @override: false, producerId: job.Id).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            Error?.Invoke(new BullMQException(
                $"Failed to add repeatable job for next iteration: {ex.Message}", ex));
        }
    }

    private async Task ProcessJob(IQueueBackend backend, Job job, string token)
    {
        var ts = job.ProcessedOn ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        _lockManager?.TrackJob(job.Id!, token, ts);
        Active?.Invoke(job);
        try
        {
            var result = await _processor(job, _abortCts.Token).ConfigureAwait(false);
            await job.MoveToCompletedAsync(result, token, fetchNext: false).ConfigureAwait(false);
            Completed?.Invoke(job, result);
        }
        catch (Exception ex)
        {
            try
            {
                await job.MoveToFailedAsync(ex, token, fetchNext: false).ConfigureAwait(false);
            }
            catch (Exception moveError)
            {
                Error?.Invoke(moveError);
            }

            Failed?.Invoke(job, ex);
        }
        finally
        {
            _lockManager?.UntrackJob(job.Id!);
        }
    }

    /// <summary>Blocks on the backend's wait primitive until a job may be available.</summary>
    private async Task WaitForJobAsync(IQueueBackend backend, long delayUntil = 0)
    {
        var maxBlockSeconds = Math.Min(Math.Max(Opts.DrainDelay, 1), 10);

        // When the next job is a delayed job due at `delayUntil`, sleep until it
        // is due (bounded) rather than blocking indefinitely on the marker.
        if (delayUntil > 0)
        {
            var delta = delayUntil - DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            if (delta <= 0)
            {
                return;
            }

            await SafeDelayAsync((int)Math.Min(delta, maxBlockSeconds * 1000L)).ConfigureAwait(false);
            return;
        }

        try
        {
            var waitTask = backend.WaitForJobAsync(maxBlockSeconds);

            // Race the (server-side) blocking wait against the close signal so a
            // closing worker does not have to wait for the block timeout. An
            // abandoned wait completes harmlessly in the background.
            var cancelTask = Task.Delay(Timeout.Infinite, _closeCts.Token);
            var done = await Task.WhenAny(waitTask, cancelTask).ConfigureAwait(false);
            if (done == waitTask)
            {
                await waitTask.ConfigureAwait(false);
            }
        }
        catch (OperationCanceledException)
        {
            // Worker is closing.
        }
        catch (Exception ex)
        {
            if (!_closing)
            {
                Error?.Invoke(ex);
            }

            await SafeDelayAsync(200).ConfigureAwait(false);
        }
    }

    private async Task StalledCheckerLoopAsync(IQueueBackend backend)
    {
        while (!_closing)
        {
            try
            {
                var stalled = await backend
                    .MoveStalledJobsToWaitAsync(Opts.MaxStalledCount, Opts.StalledInterval)
                    .ConfigureAwait(false);
                foreach (var jobId in stalled)
                {
                    Stalled?.Invoke(jobId);
                }
            }
            catch (Exception ex)
            {
                if (!_closing)
                {
                    Error?.Invoke(ex);
                }
            }

            try
            {
                await Task.Delay(Opts.StalledInterval, _closeCts.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                break;
            }
        }
    }

    private async Task SafeDelayAsync(int ms)
    {
        try
        {
            await Task.Delay(ms, _closeCts.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            // Worker is closing.
        }
    }

    /// <summary>Closes the worker, waiting for in-flight work to settle.</summary>
    public async Task CloseAsync(bool force = false)
    {
        if (_closing)
        {
            return;
        }

        _closing = true;
        _forceClosing = force;

        // Interrupt the fetcher's blocking waits. On force close also cancel the
        // processor abort signal so cooperating processors stop promptly.
        _closeCts.Cancel();
        if (force)
        {
            _abortCts.Cancel();
        }

        if (_running is not null)
        {
            try
            {
                await _running.ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                // Expected on cancellation.
            }
        }

        if (_stalledChecker is not null)
        {
            try
            {
                await _stalledChecker.ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                // Expected.
            }
        }

        if (_lockManager is not null)
        {
            await _lockManager.CloseAsync().ConfigureAwait(false);
        }

        if (_backend.IsValueCreated)
        {
            var backend = await _backend.Value.ConfigureAwait(false);
            await backend.CloseAsync(force).ConfigureAwait(false);
        }

        _closeCts.Dispose();
        _abortCts.Dispose();
        _slots?.Dispose();
    }

    public async ValueTask DisposeAsync() => await CloseAsync().ConfigureAwait(false);
}
